### PR TITLE
Enable User assigned managed identity for scalesets

### PR DESCRIPTION
**Pull request by chkeita**
## Summary of the Pull Request

Create a User Assigned Managed Identity at deplyoment 
Assign the idenity to scalesets deployed by the service

## PR Checklist
* [x] Applies to work item: #209 
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/onefuzz) and sign the CLI.
* [ ] Tests added/passed
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_

## Validation Steps Performed

- Deployed
- Enabled "User assignment required" on the entreprise application of the instance
- Verified that job can run to completion

### CODE REVIEW

File: src/api-service/__app__/onefuzzlib/azure/creds.py [hunk 1]
```diff
@@ -123,3 +124,22 @@ def is_member_of(group_id: str, member_id: str) -> bool:
             CheckGroupMembershipParameters(group_id=group_id, member_id=member_id)
         ).value
     )
+
+
+def get_scaleset_idenity_resource_path() -> str:
```
Comment (bmc-msft) [inline src/api-service/__app__/onefuzzlib/azure/creds.py:129]: If `get_scaleset_principal_id` is safe to cache, we should cache this too.

**Comment by bmc-msft** [comment]
Running integration tests

**Comment by bmc-msft** [comment]
Integration tests pass

**Review by bmc-msft — APPROVED** [review]

File: src/api-service/__app__/onefuzzlib/agent_authorization.py [hunk 2]
```diff
@@ -54,6 +55,8 @@ def try_get_token_auth_header(request: func.HttpRequest) -> Union[Error, TokenDa
 
 @cached(ttl=60)
 def is_authorized(token_data: TokenData) -> bool:
+    if get_scaleset_principal_id() == token_data.object_id:
```
Comment (ranweiler) [inline src/api-service/__app__/onefuzzlib/agent_authorization.py:58]: Let's add a comment describing what situation this is checking, at a higher level. Is this an optimization? Or does the check below miss something?

File: src/api-service/__app__/onefuzzlib/agent_authorization.py [hunk 3]
```diff
@@ -54,6 +55,8 @@ def try_get_token_auth_header(request: func.HttpRequest) -> Union[Error, TokenDa
 
 @cached(ttl=60)
 def is_authorized(token_data: TokenData) -> bool:
```
Comment (ranweiler) [inline src/api-service/__app__/onefuzzlib/agent_authorization.py:57]: Not new to this PR, but let's add a comment emphasizing why we trust the `object_id` in the `TokenData`.

File: src/api-service/__app__/onefuzzlib/azure/creds.py (hunk 1 above)
Comment (ranweiler) [inline src/api-service/__app__/onefuzzlib/azure/creds.py:129]: Typo, should be `.._identity_..`.

Comment (ranweiler) [inline src/api-service/__app__/onefuzzlib/azure/creds.py:129]: Is there documentation we can link to describing what path spec we are trying to satisfy? Maybe something in the managed identity docs?

File: src/api-service/__app__/onefuzzlib/azure/creds.py [hunk 4]
```diff
@@ -123,3 +124,22 @@ def is_member_of(group_id: str, member_id: str) -> bool:
...
+    return "%s/Microsoft.ManagedIdentity/userAssignedIdentities/%s" % (
+        resource_group_path,
+        scaleset_id_name,
+    )
+
+
+@cached
+def get_scaleset_principal_id() -> UUID:
+    client = mgmt_client_factory(ResourceManagementClient)
+    uid = client.resources.get_by_id(get_scaleset_idenity_resource_path(), "2018-11-30")
```
Comment (ranweiler) [inline src/api-service/__app__/onefuzzlib/azure/creds.py:144]: Maybe this API version should be a constant.

Is there a specific ARM template valuable it needs to match, that we should document at its definition site?

File: src/deployment/deploy.py [hunk 5]
```diff
@@ -395,6 +396,10 @@ def deploy_template(self):
             sys.exit(1)
         self.results["deploy"] = result.properties.outputs
 
+        assign_scaleset_role(
```
Comment (ranweiler) [inline src/deployment/deploy.py:399]: How about a log message here?

File: src/deployment/registration.py [hunk 6]
```diff
@@ -298,28 +298,38 @@ def assign_scaleset_role(onefuzz_instance_name: str, scaleset_name: str):
...
         % scaleset_service_principal["id"],
-        body={
-            "principalId": scaleset_service_principal["id"],
-            "resourceId": onefuzz_service_principal["id"],
-            "appRoleId": lab_machine_role["id"],
-        },
     )
 
+    # check if the role is already assigned
+    if not seq(assignments["value"]).find(
```
Comment (ranweiler) [inline src/deployment/registration.py:319]: For readability, can we maybe hoist this into a variable like:
```python
role_assigned = seq(assignments["value"]).find(lambda a: a["appRoleId"] == managed_node_role["id"])
if not role_assigned:
    query_microsoft_graph(..)
```

File: src/api-service/__app__/onefuzzlib/azure/creds.py (hunk 1 above)
Comment (ranweiler) [reply src/api-service/__app__/onefuzzlib/azure/creds.py:129]: Generally: I think relating this code to the ARM template and what we're doing conceptually will be useful in the future.

**Review by ranweiler — APPROVED** [review]

File: src/api-service/__app__/onefuzzlib/agent_authorization.py (hunk 3 above)
Comment (chkeita) [reply src/api-service/__app__/onefuzzlib/agent_authorization.py:57]: Already have a comment on line 50 for that

**Pull request merged**